### ISSUE TITLE
Cant insert new node in binary tree.

### ISSUE THREAD

**Issue by HarsheetKakar** [codezonediitj/pydatastructs#180]
#### Description of the problem

While trying to add RedBlackTree in binary_tree.py, I encountered the error:
TypeError: __new__() missing 1 required positional argument: 'data'

class BinaryTree:
```
def __new__(cls, key=None, root_data=None, comp=None,
                is_order_statistic=False, Node=RedBlackTreeNode):
        obj = object.__new__(cls)
        if key is None and root_data is not None:
            raise ValueError('Key required.')
        key = None if root_data is None else key
        root = Node(key, root_data)
        root.is_root = True
        obj.root_idx = 0
        obj.tree, obj.size = ArrayForTrees(Node, [root]), 1
        obj.comparator = lambda key1, key2: key1 < key2 \
                        if comp is None else comp
        obj.is_order_statistic = is_order_statistic
        return obj
```
```
class RedBlackTreeNode(Node):

    __slots__ = ['key', 'data', 'left', 'right', 'is_root',
                 'height', 'parent', 'size']

    def __new__(cls, key, data):
        obj = Node.__new__(cls)
        obj.data, obj.key = data, key
        obj.left, obj.right, obj.parent, obj.height, obj.size = \
            None, None, None, 0, 1
        obj.is_root = False
        return obj

    def __str__(self):
        """
        Used for printing.
        """
        return str((self.left, self.key, self.data, self.right))
```

#### Example of the problem

Steps to produce:
- Extend TreeNode class.
- Use that child class as node in BinaryTree.__new__().
- start tests.

#### References/Other comments
Error:
```
{
	"resource": "/D:/projects/pydatastructs/pydatastructs/trees/tests/test_binary_trees.py",
	"owner": "pytest",
	"code": "./pydatastructs/trees/tests/test_binary_trees.py::test_BinarySearchTree",
	"severity": 8,
	"message": "Fail: TypeError: __new__() missing 1 required positional argument: 'data'",
	"source": "pytest",
	"startLineNumber": 8,
	"startColumn": 5,
	"endLineNumber": 8,
	"endColumn": 26,
	"relatedInformation": [
		{
			"startLineNumber": 8,
			"startColumn": 5,
			"endLineNumber": 8,
			"endColumn": 26,
			"message": "test_BinarySearchTree",
			"resource": "/d:/projects/pydatastructs/pydatastructs/trees/tests/test_binary_trees.py"
		},
		{
			"startLineNumber": 13,
			"startColumn": 5,
			"endLineNumber": 13,
			"endColumn": 19,
			"message": "b.insert(3, 3)",
			"resource": "/d:/projects/pydatastructs/pydatastructs/trees/tests/test_binary_trees.py"
		},
		{
			"startLineNumber": 233,
			"startColumn": 21,
			"endLineNumber": 233,
			"endColumn": 47,
			"message": "self.tree.append(new_node)",
			"resource": "/d:/projects/pydatastructs/pydatastructs/trees/binary_trees.py"
		},
		{
			"startLineNumber": 233,
			"startColumn": 13,
			"endLineNumber": 233,
			"endColumn": 52,
			"message": "arr_new[self._last_pos_filled + 1] = el",
			"resource": "/d:/projects/pydatastructs/pydatastructs/linear_data_structures/arrays.py"
		},
		{
			"startLineNumber": 122,
			"startColumn": 17,
			"endLineNumber": 122,
			"endColumn": 41,
			"message": "elem = self._dtype(elem)",
			"resource": "/d:/projects/pydatastructs/pydatastructs/linear_data_structures/arrays.py"
		}
	]
}
```

**Comment by czgdp1807** [comment]
I cannot get what you are trying to do. However, intuitively, `RedBlackTreeNode` should inherit `TreeNode`. 
I tried the following,
```python
class RedBlackTreeNode(TreeNode):
    """
    Represents node in trees.

    Parameters
    ==========

    data
        Any valid data to be stored in the node.
    key
        Required for comparison operations.
    left: int
        Optional, index of the left child node.
    right: int
        Optional, index of the right child node.
    """

    __slots__ = ['key', 'data', 'left', 'right', 'is_root',
                 'height', 'parent', 'size']

    def __new__(cls, key, data):
        obj = Node.__new__(cls)
        obj.data, obj.key = data, key
        obj.left, obj.right, obj.parent, obj.height, obj.size = \
            None, None, None, 0, 1
        obj.is_root = False
        return obj

    def __str__(self):
        """
        Used for printing.
        """
        return str((self.left, self.key, self.data, self.right))
```
And replaced `TreeNode` with `RedBlackTreeNode`, there wasn't any problem, tests were passing.

**Comment by HarsheetKakar** [comment]
I have done exactly the same thing and it is still showing me the same error. I discarded all other changes and just retained changes in utils.__init__, utils.misc_utils.py and binary_trees.BinaryTree it is still giving me the same error

**Comment by HarsheetKakar** [comment]
![image](https://user-images.githubusercontent.com/42893005/77193983-1a642280-6b05-11ea-9354-3f0995a5b5d5.png)
this is the exact error which is popping up

**Comment by HarsheetKakar** [comment]
I just wanted to add color attribute to TreeNode so I inherited it to RedBlackTreeNode and then tried everything but the error persisted. Would you recommend any other way?

**Comment by czgdp1807** [comment]
What's the python version?

**Comment by HarsheetKakar** [comment]
3.7.3

**Comment by czgdp1807** [comment]
> I just wanted to add color attribute to TreeNode so I inherited it to RedBlackTreeNode and then tried everything but the error persisted. Would you recommend any other way?

I get your point. You may need override the `__new__` method in `RedBlackTree`. However, can we just extra `color` attribute in `TreeNode`. Default it to `None` for other trees and use that attribute in  `RedBlackTree`.

**Comment by HarsheetKakar** [comment]
We can do it this once, but it doesnt seem to be a good solution in long term.

**Comment by czgdp1807** [comment]
Well, then you will inherit `TreeNode` in `RedBlackTreeNode` and then use it in `RedBlackTree`. Another option can be is to override `BinaryTree.__new__` in `RedBlackTree.__new__`. Have you tried it?

**Comment by czgdp1807** [comment]
Can you make a PR for `RedBlackTree` using the approach which is giving the error you have described in the first comment? Let's see how it works out in CI.

**Comment by HarsheetKakar** [comment]
> Well, then you will inherit `TreeNode` in `RedBlackTreeNode` and then use it in `RedBlackTree`. Another option can be is to override `BinaryTree.__new__` in `RedBlackTree.__new__`. Have you tried it?

let me first try overriding `__new__` and if it doesnt work then I will simply add color in keyword and make an issue for someone else to resolve

**Comment by HarsheetKakar** [comment]
> Can you make a PR for `RedBlackTree` using the approach which is giving the error you have described in the first comment? Let's see how it works out in CI.

sure

**Comment by HarsheetKakar** [comment]
Not working with Travis CI either.
#181

**Comment by HarsheetKakar** [comment]
solution found, implemented it in #181
